CIP-0071/README.md

```diff
@@ -0,0 +1,246 @@
+
+#### "Ballot Box" -> Smart Contract
+
+Essentially, the "ballot box" is a smart contract with arbitrary logic decided upon by the authorized vote counter.  Some examples include:
+
+1. A ballot box that can be redeemed at any time by a tx signed by the authorized vote counter
+2. A ballot box that can be redeemed only after polls close
+2. A ballot box that can be redeemed once a majority of voters have sent in a ballot
+3. A ballot box that can be redeemed only by the specific wallet specified in the `voter` datum of each UTxO
+
+Because the ballot creation and vote casting process has already occurred on-chain we want to provide the maximum flexibility in the protocol here so that each project can decide what is best for their own community.  Helios code for the simple case enumerated as #1 above would be:
+
+```ts
+const EXPECTED_SIGNER: PubKeyHash = PubKeyHash::new(#0123456789abcdef)
+
+func main(ctx: ScriptContext) -> Bool {
+  ctx.tx.is_signed_by(EXPECTED_SIGNER)
+}
+```
+
+### The Vote Counting Process
+
+#### "Ballot Counter" -> Authorized Wallet
+
+Given the flexible nature of the ["ballot box" smart contract](#Ballot-Box---Smart-Contract) enumerated above, we propose a simple algorithm for counting votes and returning the ballots to the user:
+
+1. Ensure the polls are closed (can be either on or off-chain)
+2. Iterate through all UTxOs in forward-time-order locked in the "ballot box" and for each:
+	* Determine which assets are inside of that UTxO
+	* Mark their most recent vote to match the `vote` object in the UTxOs datum
+3. Ensure any required quorums or thresholds were reached
+4. Report on the final ballot outcome
+
+Javascript-like pseudocode using the [Lucid library](https://github.com/spacebudz/lucid) for the above algorithm would be as follows:
+
+```js
+function countVotes(ballotPolicyId, ballotBox) {
+  var votesByAsset = {};
+  const votes = await lucid.utxosAt(ballotBox);
+  for (const vote of votes) {
+    const voteResult = Data.toJson(Data.from(vote.datum));
+    for (const unit in vote.assets) {
+      if (!unit.startsWith(ballotPolicyId)) {
+        continue;
+      }
+      const voteCount = Number(vote.assets[unit]);         // Should always be 1
+      votesByAsset[unit] = {
+        voter: voteResult.voter,
+        vote: voteResult.vote,
+        count: voteCount
+      }
+    }
+  }
+  return votesByAsset;
+}
+```
+
+There is no requirement that the "ballot counter" redeem all "ballots" from the "ballot box" and send them back to the respective voters, but we anticipate that this is what will happen in practice.  We encourage further open-sourced code versions that enforce this requirement at the smart contract level.
```

Comment:
I believe it makes more sense to burn the "ballots" instead of returning them to their respective voters. There are a few reasons for this:
First, in the motivation section, you discuss the issues with governance tokens. Although the ballots are no longer usable (defined by the time in the policy), it seems unnecessarily confusing for users to receive what seems like a governance token. That could result in confusion and even scams. 
Secondly, burning the asset will make it much easier for third parties to tabulate votes or at on the results. While all the data is provided to tabulate votes as a third party already, it is quite simple, and already done for other applications, to track the -1 mint transactions of all assets under a policy ID. Any on chain voting should allow vote results to be verified as simply as possible, which would be done through burning the assets.

Of course, once you burn the asset, there is the minUTXO that was previously attached to the asset. It is important to do SOMETHING with that minUTXO, so we don't end up with up to ~10.5K ADA locked in the contract address. I propose that it is all sent back to the voters, so the only cost associated with voting is the transaction fee, but I would also support splitting it with the developer(s) in some capacity to pay for their time.

Reply:
I think the ideal here is to have the default ballot box perform:

* Burn ballot
* Send minUTxO back to address indicated by datum
* Only do the above after verifying on chain that polls are closed

I would still like to leave this as optional if possible, though it will definitely be an initial implementation in the [voting.js](https://github.com/thaddeusdiamond/cardano-nft-mint-frontend/blob/master/src/js/pages/voting.js) file in cardano-nft-mint-frontend.

Reply:
> I would still like to leave this as optional if possible, though it will definitely be an initial implementation in the [voting.js](https://github.com/thaddeusdiamond/cardano-nft-mint-frontend/blob/master/src/js/pages/voting.js) file in cardano-nft-mint-frontend.

I disagree with this. I think that if there is going to be a standard, it should be just that -- standard. A 3rd party should not have to change their tabulation system in order to verify votes. Ballots should always be burned once the polls have closed. If a voting organizer wants to give out a commemorative voting NFT (think the "I voted" sticker), they should then mint assets on a new policy ID and send those to the voters.

Leaving it up to choice in this situation seems like the worst choice, it removes the benefits of standardizing this.

Reply:
As we discussed in the chat, because the minting of the "ballot" HAS TO occur in the same transaction as the "send-to-ballot-box" to prevent a "mint now, vote later" attack, I think this is less of a concern.  Monitoring for a burn transaction would actually duplicate the cost to viewers who would have to find the burn transaction per the standard.

Reply:
Upon further discussion with @thaddeusdiamond in a Discord server (and re-reading the proposal), I now tend to agree that the burning shouldn't be required, at least not for the sake of a developer. I still do believe that burning the token makes the most sense (release the minUTXO, prevent confusion on the side of users), but it certainly isn't mandatory.

Reply:
I am not sure about it being optional or not but I do agree that burning the token makes the most sense instead of sending it back to the voter.

Reply:
> I believe it makes more sense to burn the "ballots" instead of returning them to their respective voters. 

Yes. 

>  I think that if there is going to be a standard, it should be just that -- standard.

Also.

> Upon further discussion with @thaddeusdiamond in a Discord server (and re-reading the proposal)

I'd love to see the outcome of that discussion in the 'Rationale'. In particular, the rationale for burning/not burning token needs to be stated clearly as this is clearly a point everyone seems to be tripping on.

Reply:
It's mostly about the use case around commemorative NFTs.  Will add a Rationale section with our offline discussion.